File: Groove.SP/groove-sp-web/src/app/features/cruise-order/models/cruise-order.model.ts
import { Model, ModelBase, DataType } from 'src/app/core/models/model-base.model';
import { CruiseOrderItemModel } from './cruise-order-item.model';

@Model
export class CruiseOrderModel extends ModelBase {
    id: number = 0;
    actualDeliveryDate: string | null = undefined;
    actualShipDate: string | null = undefined;
    approvalStatus: string = undefined;
    approvedDate: string | null = undefined;
    approver: string = undefined;
    budgetAccount: string = undefined;
    budgetId: string = undefined;
    budgetPeriod: number | null = undefined;
    budgetYear: number | null = undefined;
    certificateId: string = undefined;
    certificateNumber: string = undefined;
    creationUser: string = undefined;
    delivered: string = undefined;
    deliveryMeans: string = undefined;
    department: string = undefined;
    firstReceivingPoint: string = undefined;
    invoiced: number | null = undefined;
    maintenanceObject: string = undefined;
    maker: string = undefined;
    poCause: string = undefined;
    poId: string = undefined;
    poType: string = undefined;
    poPriority: string = undefined;
    requestApprovedDate: string | null = undefined;
    requestDate: string | null = undefined;
    requestPriority: string = undefined;
    requestType: string = undefined;
    requestType2: string = undefined;
    requestType3: string = undefined;
    requestor: string = undefined;
    ship: string = undefined;
    withWO: string = undefined;
    estimatedDeliveryDate: string | null = undefined;
    poNumber: string = undefined;
    poStatus: string | null = undefined;
    poSubject: string = undefined;
    poDate: string | null = undefined;
    statusName: string = undefined;
    customer: string = undefined;
    supplier: string = undefined;
    consignee: string = undefined;
    contacts: Array<any> = undefined;
    @DataType(CruiseOrderItemModel)
    items: Array<CruiseOrderItemModel> = [];
}

File: Groove.SP/Groove.SP.Core/Entities/Cruise/CruiseOrderModel.cs
using Groove.SP.Core.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Groove.SP.Core.Entities.Cruise
{
    public class CruiseOrderModel : Entity
    {
        public long Id { get; set; }
        public DateTime? ActualDeliveryDate { get; set; }
        public DateTime? ActualShipDate { get; set; }
        public string ApprovalStatus { get; set; }
        public DateTime? ApprovedDate { get; set; }
        public string Approver { get; set; }
        public string BudgetAccount { get; set; }
        public string BudgetId { get; set; }
        public int? BudgetPeriod { get; set; }
        public int? BudgetYear { get; set; }
        public string CertificateId { get; set; }
        public string CertificateNumber { get; set; }
        public string CreationUser { get; set; }
        public string Delivered { get; set; }
        public string DeliveryMeans { get; set; }
        public string Department { get; set; }
        public string FirstReceivingPoint { get; set; }
        public decimal? Invoiced { get; set; }
        public string MaintenanceObject { get; set; }
        public string Maker { get; set; }
        public string POCause { get; set; }
        public string POId { get; set; }
        public string POType { get; set; }
        public string POPriority { get; set; }
        public DateTime? RequestApprovedDate { get; set; }
        public DateTime? RequestDate { get; set; }
        public string RequestPriority { get; set; }
        public string RequestType { get; set; }
        public string RequestType2 { get; set; }
        public string RequestType3 { get; set; }
        public string Requestor { get; set; }
        public string Ship { get; set; }
        public string WithWO { get; set; }
        public DateTime? EstimatedDeliveryDate { get; set; }
        public string PONumber { get; set; }
        public CruiseOrderStatus POStatus { get; set; }
        public string POSubject { get; set; }
        public DateTime? PODate { get; set; }       

        public virtual ICollection<CruiseOrderContactModel> Contacts { get; set; }
        public virtual ICollection<CruiseOrderItemModel> Items { get; set; }

        protected override void AuditChildren(string user = AppConstant.SYSTEM_USERNAME)
        {
           
            if (Contacts != null && Contacts.Any())
            {
                foreach (var contact in Contacts)
                {
                    var utcNow = DateTime.UtcNow;
                    if (string.IsNullOrWhiteSpace(contact.CreatedBy))
                    {
                        contact.CreatedBy = user;
                        contact.CreatedDate = utcNow;
                    }

                    contact.UpdatedBy = user;
                    contact.UpdatedDate = utcNow;
                }
            }

            if (Items != null && Items.Any())
            {                
                foreach (var item in Items)
                {
                    var utcNow = DateTime.UtcNow;
                    if (string.IsNullOrWhiteSpace(item.CreatedBy))
                    {
                        item.CreatedBy = user;
                        item.CreatedDate = utcNow;
                    }

                    item.UpdatedBy = user;
                    item.UpdatedDate = utcNow;
                }
            }
        }
    }
}
